File: go-node-lambda/src/lambda.ts
import {
  GetObjectCommand,
  PutObjectCommand,
  S3Client
} from '@aws-sdk/client-s3';
import { APIGatewayProxyHandler } from 'aws-lambda';
import { Readable } from 'stream';
import { createGzip } from 'zlib';

const s3 = new S3Client({});

export const handler: APIGatewayProxyHandler = async (event) => {
  try {
    // (1) -> Read file.
    const { Body } = await s3.send(
      new GetObjectCommand({
        Bucket: process.env.BUCKET_NAME!, // Make sure it is defined.
        Key: 'file.txt'
      })
    );

    // (2) -> Compress file.
    const gzip = createGzip()
      .on('error', (err) => {
        throw err;
      })
      .on('data', (chunk) => {
        console.log('Compressed chunk size: ', chunk.length);
      });

    (Body as Readable).pipe(gzip);

    // (3) -> Upload file.
    await s3.send(
      new PutObjectCommand({
        Bucket: process.env.BUCKET_NAME!,
        Key: 'node/file.txt.gz',
        Body: gzip
      })
    );

    // (4) -> Return response to user.
    return {
      statusCode: 200,
      body: 'File uploaded successfully'
    };
  } catch (err) {
    return {
      statusCode: 500,
      body: 'Something went wrong'
    };
  }
};

File: go-node-lambda/backend/src/go/lambda.go
package main

import (
	"context"
	"os"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func errorResponse(err error) events.APIGatewayProxyResponse {
	return events.APIGatewayProxyResponse{
		Body:       "Something went wrong",
		StatusCode: 500,
	}
}

func HandleRequest() events.APIGatewayProxyResponse {
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		return errorResponse(err)
	}

	s3Client := s3.NewFromConfig(cfg)

	objectOutput, err := s3Client.GetObject(context.TODO(), &s3.GetObjectInput{
		Bucket: aws.String(os.Getenv("BUCKET_NAME")),
		Key:    aws.String("file.txt"),
	})

	if err != nil {
		return errorResponse(err)
	}
	defer objectOutput.Body.Close()

	return events.APIGatewayProxyResponse{
		Body:       "File uploaded successfully",
		StatusCode: 200,
	}
}

func main() {
	lambda.Start(HandleRequest)
}
